core/blockchain.go

```diff
@@ -1261,6 +1286,16 @@ func (bc *BlockChain) WriteBlockWithState(
 		}
 	}
 
+	if bc.cacheConfig.Disabled { // archival node
+		laggedNumber := currentBlock.NumberU64() - ClearValidatorWrappersLag
```

Comment:
Based on soph's data, it takes roughly 100ms to clear one block, which is a lot for continuous pruning every block. One optimization is to check whether the pruned block has any staking txn, or whether the block is the block reward distribution block (every 64 blocks), if none is true, then no pruning is needed (no validator wrapper change)